sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1422,11 +1422,26 @@ class Analyzer(
           resolveSubQuery(s, plans)(ScalarSubquery(_, _, exprId))
         case e @ Exists(sub, _, exprId) if !sub.resolved =>
           resolveSubQuery(e, plans)(Exists(_, _, exprId))
-        case In(value, Seq(l @ ListQuery(sub, _, exprId, _))) if value.resolved && !l.resolved =>
+        case In(values, Seq(l @ ListQuery(_, _, exprId, _)))
+            if values.forall(_.resolved) && !l.resolved =>
           val expr = resolveSubQuery(l, plans)((plan, exprs) => {
             ListQuery(plan, exprs, exprId, plan.output)
           })
-          In(value, Seq(expr))
+          val subqueryOutput = expr.plan.output
+          val resolvedIn = In(values, Seq(expr))
+          if (values.length != subqueryOutput.length) {
+            throw new AnalysisException(
```

Comment:
@mgaido91 We have this check in checkInputDataTypes and here ? Is there a way we can have the number of input check in one place ?

Reply:
@mgaido91 I quickly tried the error case to check out the message - 
```
spark-sql> select * from ut1 where (c1, c2) in (select (c1, c2) from ut2);
Error in query: Cannot analyze (named_struct('c1', ut1.`c1`, 'c2', ut1.`c2`) IN (listquery())).
The number of columns in the left hand side of an IN subquery does not match the
number of columns in the output of subquery.
#columns in left hand side: 2.
#columns in right hand side: 1.
Left side columns:
[ut1.`c1`, ut1.`c2`].
Right side columns:
[`named_struct(c1, c1, c2, c2)`].;
```
The right hand side columns looks confusing. Should we only display the value exprs or the name exprs instead of both ?

Reply:
@mgaido91 I tried the following -
```
create table in(c1 int);
insert into in values(1);
spark-sql> select * from in where c1 in (1);
1
spark-sql> select * from in where c1 in (cast(1 as float));
1
spark-sql> select * from in where (c1,c1) in ((cast(1 as float), 1));
Error in query: cannot resolve '(named_struct('c1', in.`c1`, 'c1', in.`c1`) IN (named_struct('col1', CAST(1 AS FLOAT), 'col2', 1)))' due to data type mismatch: Arguments must be same type but were: struct<c1:int,c1:int> != struct<col1:float,col2:int>; line 1 pos 31;
```
Just a question, should we have been able to do a type promotion here ?

Reply:
Overall it looks good to me. Just a few minor comments. cc @gatorsmile